README.md
<p align="center">
   <img title="Flutterwave" height="200" src="https://flutterwave.com/images/logo-colored.svg" width="50%"/>
</p>

# Flutterwave Flutter SDK

## Table of Contents

- [About](#about)
- [Getting Started](#getting-started)
- [Usage](#usage)
- [Deployment](#deployment)
- [Built Using](#build-tools)
- [References](#references)
- [Support](#support)

<a id="about"></a>
## About
Flutterwave's Flutter SDK is Flutterwave's offical flutter sdk to integrate the Flutterwave payment into your flutter app. It comes with a readymade Drop In UI.
The payment methods currently supported are Cards, USSD, Mpesa, GH Mobile Money, UG Mobile Money, ZM Mobile Money, Rwanda Mobile Money, Franc Mobile Money and Nigeria Bank Account.


<a id="getting-started"></a>

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See [deployment](#deployment) for notes on how to deploy the project on a live system.
See [references](#references) for links to dashboard and API documentation.

### Prerequisite

- Ensure you have your test (and live) [API keys](https://developer.flutterwave.com/docs/api-keys).
- This Library runs and the V3 API, and so you **need** to be PCI-DSS certified to use this.
```
Flutter version >= 1.17.0
Flutterwave version 3 API keys
```

 ### Installing
Add the dependency

In your `pubspec.yaml` file add:

1. `flutterwave: 0.0.3-dev.3`
2. run `flutter pub get`

<a id="usage"></a>
## Usage

### 1. Create a `Flutterwave` instance

Create a `Flutterwave` instance by calling the constructor `Flutterwave.forUIPayment()` The constructor accepts a mandatory instance of the following:
 the calling `Context` , `publicKey`, `encryptionKey`, `amount`, `currency`, `email`, `fullName`, `txRef`, `isDebugMode` and `phoneNumber` . It returns an instance of `Flutterwave`  which we then call the `async` method `.initializeForUiPayments()` on.

     beginPayment async () { 
       try { 
		     Flutterwave flutterwave = Flutterwave.forUIPayment(
                         context: this.context,
                         encryptionKey: "Your_test_encryption_key_here",
                         publicKey: "Your_test_public_key_here",
                         currency: currency,
                         amount: amount,
                         email: "valid@email.com",
                         fullName: "Valid Full Name",
                         txRef: txref,
                         isDebugMode: true,
                         phoneNumber: "0123456789",
                         acceptCardPayment: true,
                         acceptUSSDPayment: false,
                         acceptAccountPayment: false,
                         acceptFrancophoneMobileMoney: false,
                         acceptGhanaPayment: false,
                         acceptMpesaPayment: false,
                         acceptRwandaMoneyPayment: true,
                         acceptUgandaPayment: false,
                         acceptZambiaPayment: false)
                         
             final ChargeResponse response = await flutterwave.initializeForUiPayments();
                        
	         } catch(error) {
		         handleError(error);
	         }



### 2. Handle the response

 Calling the `.initialiazeForUiPayments()` method returns a `Future`
 of `ChargeResponse` which we await for the actual response as seen above.
 
 An example of how to make payment in a Widget would look like this:
 
 ```
    class PaymentWidget extends StatefulWidget {
      @override
      _PaymentWidgetState createState() => _PaymentWidgetState();
    }
    
    class _PaymentWidgetState extends State<PaymentWidget> {
      final String txref = "My_unique_transaction_reference_123";
      final String amount = "200";
      final String currency = FlutterwaveCurrency.RWF;
    
      @override
      Widget build(BuildContext context) {
        return Container();
      }
    
      beginPayment() async {
        final Flutterwave flutterwave = Flutterwave.forUIPayment(
            context: this.context,
            encryptionKey: "FLWPUBK_TEST-SANDBOXDEMOKEY-X",
            publicKey: "FLWPUBK_TEST-SANDBOXDEMOKEY-X",
            currency: this.currency,
            amount: this.amount,
            email: "valid@email.com",
            fullName: "Valid Full Name",
            txRef: this.txref,
            isDebugMode: true,
            phoneNumber: "0123456789",
            acceptCardPayment: true,
            acceptUSSDPayment: false,
            acceptAccountPayment: false,
            acceptFrancophoneMobileMoney: false,
            acceptGhanaPayment: false,
            acceptMpesaPayment: false,
            acceptRwandaMoneyPayment: true,
            acceptUgandaPayment: false,
            acceptZambiaPayment: false);
    
        try {
          final ChargeResponse response = await flutterwave.initializeForUiPayments();
          if (response == null) {
            // user didn't complete the transaction.
          } else {
            final isSuccessful = checkPaymentIsSuccessful(response);
            if (isSuccessful) {
              // provide value to customer
            } else {
              // check message
              print(response.message);
    
              // check status
              print(response.status);
    
              // check processor error
              print(response.data.processorResponse);
            }
          }
        } catch (error, stacktrace) {
          // handleError(error);
        }
      }
    
      bool checkPaymentIsSuccessful(final ChargeResponse response) {
        return response.data.status == FlutterwaveConstants.SUCCESSFUL &&
            response.data.currency == this.currency &&
            response.data.amount == this.amount &&
            response.data.txRef == this.txref;
      }
    }
```
 
#### Please note that:
 - `ChargeResponse` can be null, depending on if the user cancels
   the transaction by pressing back.
 - You need to check the status of the transaction from the instance of `ChargeResponse` returned from calling `.initializeForUiPayments()`, the `amount`, `currency` and `txRef` are correct before providing value to the customer
 - To accept payment of different kinds, you need set the currency to the correspending payment type i.e, `KES` for `Mpesa`, `RWF` for `Rwanda Mobile Money`, `NGN` for `USSD`,
`Bank Accounts Payment` and so on.

>  **PLEASE NOTE**

> We advise you to do a further verification of transaction's details on your server to be sure everything checks out before providing service or goods as seen in the `checkPaymentIsSuccessful()` method above.

<a id="deployment"></a>
## Deployment

- Switch to Live Mode on the Dashboard settings page
- Use the Live Public API key from the API tab, see [here](https://developer.flutterwave.com/docs/api-keys) for more details.

<a id="build-tools"></a>
## Built Using
- [flutter](https://flutter.dev/)
- [dart](https://dart.dev/)
- [http](https://pub.dev/packages/http)
- [dart_des](https://pub.dev/packages/dart_des)
- [webview_flutter](https://pub.dev/packages/webview_flutter)

<a id="references"></a>
## Flutterwave API  References

- [Flutterwave API Doc](https://developer.flutterwave.com/docs)
- [Flutterwave Inline Payment Doc](https://developer.flutterwave.com/docs/flutterwave-inline)
- [Flutterwave Dashboard](https://dashboard.flutterwave.com/login)  

<a id="support"></a>
## Support
 - Compatibility has been bumped up to Android minSdk 19
 - Supporting Android 4.4+ for reference checkout [Android 4.4 (API level 19)](https://developer.android.com/studio/releases/platforms#4.4)
* Have issues integrating? Reach out via [our Developer forum](https://developer.flutterwave.com/discuss) for support


CHANGELOG.md
## [0.1.0] - Jun 21, 2021
* Null safe implementation
* New better encryption

## [0.0.3-dev.3] - May 19, 2021
* Added support for card payment in ZAR

## [0.0.3-dev.2] - March 12, 2021
* Added support for NO_AUTH cards

## [0.0.3-dev.1]
* Optimized payment home screen - March 11, 2021
* Displayed debug label only on production
* Added metric tracking
* Updated UI
* Added Bank Transfer Payment
* Updated dependencies

## [0.0.2] 
* Updated dependencies project dependencies add example project - October 9, 2020
* Added example project

## [0.0.1] 
* Intital release of Flutterwave's flutter library - October 9, 2020


lib/widgets/home/flutterwave_payment.dart
import 'package:flutter/material.dart';
import 'package:flutterwave/core/bank_transfer_manager/bank_transfer_payment_manager.dart';
import 'package:flutterwave/core/card_payment_manager/card_payment_manager.dart';
import 'package:flutterwave/core/flutterwave_payment_manager.dart';
import 'package:flutterwave/core/mobile_money/mobile_money_payment_manager.dart';
import 'package:flutterwave/core/mpesa/mpesa_payment_manager.dart';
import 'package:flutterwave/core/pay_with_account_manager/bank_account_manager.dart';
import 'package:flutterwave/core/ussd_payment_manager/ussd_manager.dart';
import 'package:flutterwave/core/voucher_payment/voucher_payment_manager.dart';
import 'package:flutterwave/models/responses/charge_response.dart';
import 'package:flutterwave/widgets/bank_account_payment/bank_account_payment.dart';
import 'package:flutterwave/widgets/bank_transfer_payment/bank_transfer_payment.dart';
import 'package:flutterwave/widgets/card_payment/card_payment.dart';
import 'package:flutterwave/widgets/mobile_money/pay_with_mobile_money.dart';
import 'package:flutterwave/widgets/mpesa_payment/pay_with_mpesa.dart';
import 'package:flutterwave/widgets/ussd_payment/pay_with_ussd.dart';
import 'package:flutterwave/widgets/voucher_payment/pay_with_voucher.dart';

import 'flutterwave_payment_option.dart';

class FlutterwaveUI extends StatefulWidget {
  final FlutterwavePaymentManager _flutterwavePaymentManager;

  FlutterwaveUI(this._flutterwavePaymentManager);

  @override
  _FlutterwaveUIState createState() => _FlutterwaveUIState();
}

class _FlutterwaveUIState extends State<FlutterwaveUI> {
  final GlobalKey<ScaffoldState> _scaffoldKey = GlobalKey<ScaffoldState>();

  @override
  Widget build(BuildContext context) {
    final FlutterwavePaymentManager paymentManager =
        this.widget._flutterwavePaymentManager;

    return MaterialApp(
      debugShowCheckedModeBanner: paymentManager.isDebugMode,
      theme: ThemeData(fontFamily: "FLW"),
      home: Scaffold(
        key: this._scaffoldKey,
        body: SingleChildScrollView(
          child: ConstrainedBox(
            constraints: BoxConstraints(
              maxHeight: MediaQuery.of(context).size.height,
            ),
            child: SafeArea(
              child: Container(
                width: double.infinity,
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: [
                    Container(
                      margin: EdgeInsets.fromLTRB(10, 50, 10, 70),
                      child: Column(
                        children: [
                          Row(
                            mainAxisSize: MainAxisSize.max,
                            children: [
                              Icon(
                                Icons.lock,
                                size: 10.0,
                                color: Colors.black,
                              ),
                              SizedBox(
                                width: 5.0,
                              ),
                              Text(
                                "SECURED BY FLUTTERWAVE",
                                style: TextStyle(
                                    color: Colors.black,
                                    fontSize: 10.0,
                                    fontFamily: "FLW",
                                    letterSpacing: 1.0),
                              )
                            ],
                          ),
                          SizedBox(
                            width: double.infinity,
                            height: 100,
                          ),
                          Container(
                            width: double.infinity,
                            margin: EdgeInsets.fromLTRB(0, 100, 0, 0),
                            child: Text(
                              "How would you \nlike to pay?",
                              textAlign: TextAlign.left,
                              style: TextStyle(
                                fontWeight: FontWeight.bold,
                                color: Colors.black,
                                fontFamily: "FLW",
                                fontSize: 30.0,
                              ),
                            ),
                          ),
                          Align(
                            alignment: Alignment.centerLeft,
                            child: Container(
                              height: 5,
                              width: 200,
                              color: Colors.pink,
                              margin: EdgeInsets.fromLTRB(0, 20, 0, 10),
                            ),
                          ),
                        ],
                      ),
                    ),
                    Container(
                      color: Colors.white38,
                      width: double.infinity,
                      child: Column(
                        children: [
                          Visibility(
                            visible: paymentManager.acceptAccountPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  width: double.infinity,
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick:
                                        this._launchPayWithAccountWidget,
                                    buttonText: "Account",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptCardPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  width: double.infinity,
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick: this._launchCardPaymentWidget,
                                    buttonText: "Card",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptUSSDPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick: this._launchUSSDPaymentWidget,
                                    buttonText: "USSD",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptBankTransferPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick:
                                        this._launchBankTransferPaymentWidget,
                                    buttonText: "Bank Transfer",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptMpesaPayment,
                            child: Column(children: [
                              SizedBox(
                                height: 50.0,
                                child: FlutterwavePaymentOption(
                                  handleClick: this._launchMpesaPaymentWidget,
                                  buttonText: "Mpesa",
                                ),
                              ),
                              SizedBox(
                                height: 0.5,
                              ),
                            ]),
                          ),
                          Visibility(
                            visible: paymentManager.acceptRwandaMoneyPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick:
                                        this._launchMobileMoneyPaymentWidget,
                                    buttonText: "Rwanda Mobile Money",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptGhanaPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick:
                                        this._launchMobileMoneyPaymentWidget,
                                    buttonText: "Ghana Mobile Money",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptUgandaPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick:
                                        this._launchMobileMoneyPaymentWidget,
                                    buttonText: "Uganda Mobile Money",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptZambiaPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick:
                                        this._launchMobileMoneyPaymentWidget,
                                    buttonText: "Zambia Mobile Money",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: false,
                            child: Column(
                              children: [
                                SizedBox(
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick: () => {},
                                    buttonText: "Barter",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptVoucherPayment,
                            child: Column(
                              children: [
                                SizedBox(
                                  height: 50.0,
                                  child: FlutterwavePaymentOption(
                                    handleClick:
                                        this._launchVoucherPaymentWidget,
                                    buttonText: "Voucher",
                                  ),
                                ),
                                SizedBox(
                                  height: 0.5,
                                ),
                              ],
                            ),
                          ),
                          Visibility(
                            visible: paymentManager.acceptFancophoneMobileMoney,
                            child: SizedBox(
                              height: 50.0,
                              child: FlutterwavePaymentOption(
                                handleClick:
                                    this._launchMobileMoneyPaymentWidget,
                                buttonText: "Francophone Mobile Money",
                              ),
                            ),
                          )
                        ],
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }

  void _launchCardPaymentWidget() async {
    final CardPaymentManager cardPaymentManager =
        this.widget._flutterwavePaymentManager.getCardPaymentManager();

    final dynamic response = await Navigator.push(
      this.context,
      MaterialPageRoute(builder: (context) => CardPayment(cardPaymentManager)),
    );
    _handleBackPress(response);
  }

  void _launchPayWithAccountWidget() async {
    final BankAccountPaymentManager bankAccountPaymentManager =
        this.widget._flutterwavePaymentManager.getBankAccountPaymentManager();
    final response = await Navigator.push(
      this.context,
      MaterialPageRoute(
          builder: (context) => PayWithBankAccount(bankAccountPaymentManager)),
    );
    _handleBackPress(response);
  }

  void _launchUSSDPaymentWidget() async {
    final USSDPaymentManager paymentManager =
        this.widget._flutterwavePaymentManager.getUSSDPaymentManager();
    final response = await Navigator.push(
      this.context,
      MaterialPageRoute(builder: (context) => PayWithUssd(paymentManager)),
    );
    _handleBackPress(response);
  }

  void _launchMobileMoneyPaymentWidget() async {
    final MobileMoneyPaymentManager mobileMoneyPaymentManager =
        this.widget._flutterwavePaymentManager.getMobileMoneyPaymentManager();
    final response = await Navigator.push(
      this.context,
      MaterialPageRoute(
          builder: (context) => PayWithMobileMoney(mobileMoneyPaymentManager)),
    );
    _handleBackPress(response);
  }

  void _launchMpesaPaymentWidget() async {
    final FlutterwavePaymentManager paymentManager =
        this.widget._flutterwavePaymentManager;
    final MpesaPaymentManager mpesaPaymentManager =
        paymentManager.getMpesaPaymentManager();
    final response = await Navigator.push(
      this.context,
      MaterialPageRoute(
          builder: (context) => PayWithMpesa(mpesaPaymentManager)),
    );
    _handleBackPress(response);
  }

  void _launchVoucherPaymentWidget() async {
    final VoucherPaymentManager voucherPaymentManager =
        this.widget._flutterwavePaymentManager.getVoucherPaymentManager();
    final response = await Navigator.push(
      this.context,
      MaterialPageRoute(
          builder: (context) => PayWithVoucher(voucherPaymentManager)),
    );
    _handleBackPress(response);
  }

  void _launchBankTransferPaymentWidget() async {
    final BankTransferPaymentManager bankTransferPaymentManager =
        this.widget._flutterwavePaymentManager.getBankTransferPaymentManager();
    final response = await Navigator.push(
      this.context,
      MaterialPageRoute(
          builder: (context) =>
              PayWithBankTransfer(bankTransferPaymentManager)),
    );
    _handleBackPress(response);
  }

  void showSnackBar(String message) {
    SnackBar snackBar = SnackBar(
      content: Text(
        message,
        textAlign: TextAlign.center,
      ),
    );
    ScaffoldMessenger.of(this._scaffoldKey.currentContext!)
        .showSnackBar(snackBar);
  }

  void _handleBackPress(dynamic result) {
    if (result == null || result is ChargeResponse) {
      final ChargeResponse? chargeResponse = result as ChargeResponse;
      String message;
      if (chargeResponse != null) {
        message = chargeResponse.message!;
      } else {
        message = "Transaction cancelled";
      }
      this.showSnackBar(message);
      Navigator.pop(this.context, chargeResponse);
    } else {
      // checking if back arrow was pressed so we do nothing.
    }
  }
}
